diff --git a/libs/parser/hconf.gperf b/libs/parser/hconf.gperf
index ead02ed..e5f0af7 100644
--- a/libs/parser/hconf.gperf
+++ b/libs/parser/hconf.gperf
@@ -9,7 +9,7 @@
 
 struct kconf_id;
 
-static const struct kconf_id *kconf_id_lookup(register const char *str, register unsigned int len);
+static const struct kconf_id *kconf_id_lookup(register const char *str, register GPERF_LEN_TYPE len);
 
 %%
 mainmenu,	T_MAINMENU,	TF_COMMAND
